client/src/permissions/UserAuth.js

```diff
@@ -0,0 +1,134 @@
+import { UserRoles } from "./UserRoles";
+import "./typedefs"
+import { GlobalContext } from "./contexts/GlobalContext";
+import { TourneyContext } from "./contexts/TourneyContext";
+import { MatchContext } from "./contexts/MatchContext";
+
+export class UserAuth {
+  /**
+   * @type {User}
+   */
+  _user;
+
+  /**
+   * @param {User} user
+   */
+  constructor(user) {
+    this._user = user;
+  }
+
+  /**
+   * @returns {User}
+   */
+  getUser() {
+    return this._user;
+  }
+
+  /**
+   * @param {string} context 
+   * @returns 
+   */
+  withContext(context) {
+    return new UserAuthWithContext(this, context);
+  }
+
+  forGlobal() {
+    return this.withContext(new GlobalContext());
+  }
+
+  /**
+   * @param {string} tourney 
+   * @returns {UserAuthWithContext}
+   */
+  forTourney(tourney) {
+    return this.withContext(new TourneyContext(tourney));
+  }
+
+  /**
+   * @param {Match} match 
+   * @param {number | undefined} playerNo
+   * @param {Map<string, Team> | Team[]} teams
+   * @returns {UserAuthWithContext}
+   */
+  forMatch(match, playerNo, teams) {
+    return this.withContext(new MatchContext(match, playerNo, teams));
+  }
+
+  /**
+   * @param {Team} match 
+   * @returns {UserAuthWithContext}
+   */
+  forTeam(team) {
+    return this.withContext(new TeamContext(team));
+  }
+}
+
+export class UserAuthWithContext {
+  /**
+   * @type {UserAuth}
+   */
+  _userAuth;
+
+  /**
+   * @type {PermissionContext}
+   */
+  _context;
+
+  /**
+   * @type {User}
+   */
+  _user
+
+  /**
+   * @param {UserAuth} userAuth 
+   * @param {PermissionContext} context 
+   */
+  constructor(userAuth, context) {
+    this._userAuth = userAuth;
+    this._context = context;
+    this._user = userAuth.getUser();
+  }
+
+  /**
+   * @param {string} role 
+   */
+  hasRole(role) {
+    return this._context.hasRole(this._user, role);
+  }
+
+  /**
+   * @param {string[]} roles 
+   * @returns {boolean}
+   */
+  hasAllRoles(roles) {
+    // Do we wanna handle host/dev like this?
+    if (this.hasAnyRole([UserRoles.Host, UserRoles.Developer])) return true;
```

Comment:
we should rethink how the Host/Developer check works, since that logic is sort of duplicated between `hasAllRoles` and `hasAnyRole`, and the logic is missing from `hasRole`

brainstorming a bit, here's one idea for some structure that just moves the check to one place
```
// Tournament roles that override checks and return true for all permissions
// (maybe there's a better name for this)
const SUPER_ROLES = [UserRoles.Host, UserRoles.Developer];

export class UserAuthWithContext {
  constructor(userAuth, context) {
    ...
    this._has_super_role = SUPER_ROLES.some(role => this._context.hasRole(this._user, role);
  }
  
  hasRole(role) {
    return this._has_super_role || this._context.hasRole(this._user, role);
  }
}
```

Reply:
Applied in b243dc59ef127357bff553b0c84ea95249737ed5

In addition to `hasRole`, `hasAnyRole` and `hasAllRoles` also checks for this._hasSuperRole, to account for the case of empty role array and avoid repeated super role checking.